File: ghidradbg_backend/src/state/stack.rs
use crate::state::Value;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct StackTrace {
    frames: Vec<StackFrame>,
}

impl StackTrace {
    pub fn new(frames: Vec<StackFrame>) -> Self {
        Self { frames }
    }
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct StackFrame {
    index: u32,
    instruction_offset: u64,
    return_offset: u64,
    params: Vec<Value>,
}

impl StackFrame {
    pub fn new(
        index: u32,
        instruction_offset: u64,
        return_offset: u64,
        params: Vec<Value>,
    ) -> Self {
        Self {
            index,
            instruction_offset,
            return_offset,
            params,
        }
    }
}

File: frontends/ghidra/src/main/java/com/github/garyttierney/ghidradbg/client/state/stack.kt
package com.github.garyttierney.ghidradbg.client.state

import kotlinx.serialization.Serializable

@Serializable
data class StackFrame(val index: Int, val instruction_offset: Long, val return_offset: Long, val params: List<Value>)

@Serializable
data class StackTrace(val frames: List<StackFrame>)
